src/Compilers/CSharp/Portable/Binder/Binder_Patterns.cs

```diff
@@ -1734,100 +1734,144 @@ private BoundPattern BindBinaryPattern(
+                currentNode = currentNode.Left as BinaryPatternSyntax;
+            } while (currentNode != null);
 
-                permitDesignations = false; // prevent designators under 'or'
-                var left = BindPattern(node.Left, inputType, permitDesignations, hasErrors, diagnostics);
-                var right = BindPattern(node.Right, inputType, permitDesignations, hasErrors, diagnostics);
+            Debug.Assert(binaryPatternStack.Count > 0);
 
-                // Compute the common type. This algorithm is quadratic, but disjunctive patterns are unlikely to be huge
-                var narrowedTypeCandidates = ArrayBuilder<TypeSymbol>.GetInstance(2);
-                collectCandidates(left, narrowedTypeCandidates);
-                collectCandidates(right, narrowedTypeCandidates);
-                var narrowedType = leastSpecificType(node, narrowedTypeCandidates, diagnostics) ?? inputType;
-                narrowedTypeCandidates.Free();
+            var binaryPatternAndPermitDesignations = binaryPatternStack.Pop();
+            BoundPattern result = BindPattern(binaryPatternAndPermitDesignations.pat.Left, inputType, binaryPatternAndPermitDesignations.permitDesignations, hasErrors, diagnostics);
+            var narrowedTypeCandidates = ArrayBuilder<TypeSymbol>.GetInstance(2);
+            collectCandidates(result, narrowedTypeCandidates);
 
-                return new BoundBinaryPattern(node, disjunction: isDisjunction, left, right, inputType: inputType, narrowedType: narrowedType, hasErrors);
+            do
+            {
+                result = bindBinaryPattern(
+                    result,
+                    this,
+                    binaryPatternAndPermitDesignations,
+                    inputType,
+                    narrowedTypeCandidates,
+                    hasErrors,
+                    diagnostics);
+            } while (binaryPatternStack.TryPop(out binaryPatternAndPermitDesignations));
 
-                static void collectCandidates(BoundPattern pat, ArrayBuilder<TypeSymbol> candidates)
-                {
-                    if (pat is BoundBinaryPattern { Disjunction: true } p)
-                    {
-                        collectCandidates(p.Left, candidates);
-                        collectCandidates(p.Right, candidates);
-                    }
-                    else
-                    {
-                        candidates.Add(pat.NarrowedType);
-                    }
-                }
+            binaryPatternStack.Free();
+            narrowedTypeCandidates.Free();
+            return result;
 
-                TypeSymbol? leastSpecificType(SyntaxNode node, ArrayBuilder<TypeSymbol> candidates, BindingDiagnosticBag diagnostics)
+            static BoundPattern bindBinaryPattern(
+                BoundPattern preboundLeft,
+                Binder binder,
+                (BinaryPatternSyntax pat, bool permitDesignations) patternAndPermitDesignations,
```

Comment:
>(BinaryPatternSyntax pat, bool permitDesignations) patternAndPermitDesignations [](http://example.com/codeflow?start=16&length=79)

Consider splitting the tuple into two separate parameters.

Reply:
It was originally separate parameters; I moved it to the tuple for simplicity of the calling code.

Reply:
The method is not using the tuple as a tuple though. It feels like the caller should be responsible for deconstructing the tuple rather than the method.